rust/lance-linalg/src/simd/f16.c

```diff
@@ -54,18 +44,25 @@ float FUNC(dot_f16)(const FP16 *x, const FP16 *y, uint32_t dimension) {
 
 #pragma clang loop unroll(enable) interleave(enable) vectorize(enable)
   for (uint32_t i = 0; i < dimension; i++) {
-    sum += (float) x[i] * (float) y[i];
+    // Use float32 as the accumulator to avoid overflow.
+    sum += x[i] * y[i];
   }
   return sum;
 }
 
 float FUNC(l2_f16)(const FP16 *x, const FP16 *y, uint32_t dimension) {
-  float sum = 0.0;
+  float sum = 0;
 
 #pragma clang loop unroll(enable) interleave(enable) vectorize(enable)
   for (uint32_t i = 0; i < dimension; i++) {
-    float s = (float) x[i] - (float) y[i];
-    sum += s * s;
+#if defined(__aarch64__)
+    // on aarch64 with fp16, this is 2x faster.
+    FP16 sub = x[i] - y[i];
+#else
+    float sub = x[i] - y[i];
+#endif
+    // Use float32 as the accumulator to avoid overflow.
+    sum += sub * sub;
```

Comment:
should we just have `simd/genric`, `simd/x86` and `simd/aarch64`?

Reply:
As 3 different functions?